### ISSUE TITLE
Apparent difference in behavior between mysql and postgres for survey timestamp fields

### ISSUE THREAD

**Issue by targetpharma** [IUREDCap/redcap-etl-module#25]
We are currently using v1.3.0 of the external module.

We have a survey instrument that has a configuration that looks like the following, which attempts to pull in some of survey metadata:

```
TABLE,promis_item_pool_v10_itchseverity,root,EVENTS
FIELD,redcap_data_access_group,varchar(255)
FIELD,redcap_survey_identifier,string
FIELD,promis_item_pool_v10_itchseverity_timestamp,datetime
...
```

We primarily use mysql as the target database. And for the above instrument there are a small handful that do not have a value in the `promis_item_pool_v10_itchseverity_timestamp` field. These appear as `0000-00-00 00:00:00` in the mysql database:

![image](https://user-images.githubusercontent.com/20141935/96038576-4e9e8800-0e35-11eb-9f8f-6fa9f40b5dc4.png)

We are testing the use of postgres as the target database. For the same instrument and configuration, we get this error when running the job:

```
Database error while trying to insert values into table ""promis_item_pool_v10_itchseverity"": SQLSTATE[22007]: Invalid datetime format: 7 ERROR:  invalid input syntax for type timestamp with time zone: "[not completed]"
LINE 1: ...,'701-020','3month_arm_2','701__clear_dermato','','[not comp...
```

Is this something you have seen before? Is the difference in behavior at the database level (mysql vs postgres) or is the REDCap-ETL code doing something differently in one case vs the other.

The workaround in this case was to switch the field type when using postgres from `datetime` to `string`, which then leads to the `[not completed]` text being in the postgres database for these cases, but at least is runs without error.

![image](https://user-images.githubusercontent.com/20141935/96040035-a0e0a880-0e37-11eb-8b34-28247b0f9cbb.png)

**Comment by mullen2** [comment]
@targetpharma 

Thanks for the feedback. I have not seen this before, and, so far, I have not been able to recreate this error. I have tried creating records directly in REDCap, and using surveys, where I have left a date field blank, and I used datetime as the field type in the transformation rules. I get a NULL value stored in Postgres for the datetime field in all cases.

**Comment by targetpharma** [comment]
Thanks for looking into it. I'll find some time on our end to see if I can reproduce it in a new/isolated project. Thanks!

**Comment by targetpharma** [comment]
@mullen2 ... I think I found the use case the triggers this. It is a partially completed survey. So in this case the study participant started filling out the survey but did not complete it. Recall, this is not an explicit date field in the survey, but the metadata about when the survey was completed and is provided by the API when you set `exportSurveyFields` to `true`.

This survey presents as multiple pages. The first page:

![image](https://user-images.githubusercontent.com/20141935/96151161-1cdefd00-0ed9-11eb-92b7-400ee8945306.png)

If the first page is completed and then 'Save and Return Later' is hit, it leads to the partial response which then populates the date as `[not completed]`.

The view from REDCap after the study participant has partially completed a survey.

![image](https://user-images.githubusercontent.com/20141935/96151081-08026980-0ed9-11eb-92b5-3eee52e9337a.png)


![image](https://user-images.githubusercontent.com/20141935/96150077-0a17f880-0ed8-11eb-9add-dd7f15396eda.png)

```json
    {
        "redcap_survey_identifier": "",
        "promis_item_pool_v10_itchseverity_timestamp": "[not completed]",
        "piqseverity02": "4",
        "piqseverity04": "3",
        "piqseverity06": "2",
        "piqseverity03": "",
        "piqseverity05": "",
        "piqseverity07": "",
        "piqseverity08": "",
        "piqseverity09___early": "0",
        "piqseverity09___day": "0",
        "piqseverity09___late": "0",
        "piqseverity09___night": "0",
        "promis_item_pool_v10_itchseverity_complete": "0"
    },
```

**Comment by mullen2** [comment]
@targetpharma Thanks for the additional information. I am able to recreate the error now, and I am looking into a fix for this issue.

**Comment by mullen2** [comment]
@targetpharma In looking into this issue, I think I have figured out a way to auto-generate the survey fields in the transformation rules (something you previously requested). I am generating the survey timestamp field in each form that is enabled as a survey.  For the survey identifier field, I am currently also generating it (only) in each form that is enabled as a survey. Is that where you would expect it to be generated?

**Comment by targetpharma** [comment]
> I think I have figured out a way to auto-generate the survey fields in the transformation rules (something you previously requested). I am generating the survey timestamp field in each form that is enabled as a survey. For the survey identifier field, I am currently also generating it (only) in each form that is enabled as a survey. Is that where you would expect it to be generated?

Yes, that sounds precisely like the expected behavior.

Curious how you will be doing that, because as I recall I was not able to determine how to use the API to distinguish if a form was enabled as a survey. But I only looked for a brief while, and it wasn't obvious at the time.

**Comment by mullen2** [comment]
@targetpharma FYI, changes for this issue and an auto-generation option for survey fields have been released in the REDCap-ETL application, and we are working on a release for the  REDCap-ETL external module.

**Comment by mullen2** [comment]
@targetpharma the new version of the external module that was just released (1.4.0) should have the fix for this issue, as well as additional transformation rules auto-generation options.
